### ISSUE TITLE
[RFE] ical calendar sync, round-trip syncing of RFC-5545 icalendar

### ISSUE THREAD

**Issue by hgkamath** [200ok-ch/organice#866]
**Is your feature request related to a problem? Please describe.**
- The problem is that the ical-calendar and org-mode-todo are two disparate entities. 

**Describe the solution you'd like**
- ical calendar sync , google/nextcloud calendar sync, 
- Round-trip syncing of RFC-5545 https://datatracker.ietf.org/doc/html/rfc5545 
- This would allow one to use organice in conjunction with other softwares that use ical, such as a calendar-app, a mail client, a browser, messaging-app etc, which have triggered hooks that set events into the calendar (such as meeting event).
- As organice is a GUI app that allows for most manipulation through GUI elements, organice can be rigid and prevent screw up of org-mode entry. A user, if manually editing an org-mode entry that corresponds to an ical-entry, will be expected to exercise care.
- This feature will make org-mode/organice interoperate with other scheduling tools and perhaps draw in more users.
- One reason to be drawn to org-mode is that one's todo/schedule is in a human readable markdown like format which when backed-up or git-committed is like a source of truth.

**Describe alternatives you've considered**
- NA
- While emacs is capable of ical export, I am unsure if it can do true round tripping, not that using emacs is an option,  [emacs manual: icalendar export](https://www.gnu.org/software/emacs/manual/html_node/org/iCalendar-Export.html)

**Additional context**
-NA

**Notes**
- An iCalendar format consists of components VCALENDAR, VEVENT, VJOURNAL, VTODO, VFREEBUSY, VTIMEZONE, ALARM, STANDARD, DAYLIGHT [ref-8.3.1](https://datatracker.ietf.org/doc/html/rfc5545#section-8.3.1)
- An org-mode entry and its corresponding ical entry, should be kept track of using org-mode-entry-properties and ical-[properties](https://datatracker.ietf.org/doc/html/rfc5545#section-8.3.2).
  - The ICAL-URL to resource containing VCALENDAR object to keep track of from which ical the org-mode entry came from
  - The property VEVENT:UID (Unique ID), Each ical-object is supposed to have a UID (unique ID) [ref-3.8.4.7](https://datatracker.ietf.org/doc/html/rfc5545#section-3.8.4.7)
  - The property VEVENT:DTSTAMP property specifies the date and time that the instance of the iCalendar object was created [ref-3.8.2.](https://datatracker.ietf.org/doc/html/rfc5545#section-3.8.2.4)
  - The property *:x-prop could hold a reverse identifier into org-mode [ref-3.8.8.2](https://datatracker.ietf.org/doc/html/rfc5545#section-3.8.8.2) 
- Needs a way to precisely map an ical-object to an org-mode-entry 
  - Operations done on a preexisting org-mode entry linked to ical-entry, will need to be inferred, and its corresponding org-mode entry will need to be updated.
- Needs a way to precisely reconstruct an ical-object from an org-mode-entry. 
  - If entry is fresh/new, then a new ical-entry is to be made. 
  - If entry is not a new ical-entry, find and modify previously existing ical entry. 
- Example of an icalendar object
   ```
       BEGIN:VCALENDAR
       VERSION:2.0
       PRODID:-//hacksw/handcal//NONSGML v1.0//EN
       BEGIN:VEVENT
       UID:19970610T172345Z-AF23B2@example.com
       DTSTAMP:19970610T172345Z
       DTSTART:19970714T170000Z
       DTEND:19970715T040000Z
       SUMMARY:Bastille Day Party
       END:VEVENT
       END:VCALENDAR
  ```
- When an org-entry linked to an ical-entry has to be deleted, it will first have to marked for deletion, and purged only after it is confirmed that entry no longer exists in the ical.
- Some care will be needed for handling those operations outside the intersection of do-able operations of ical and org-mode.

**Comment by hgkamath** [comment]
Before creating a new issue, please read the "How we work with issues" section in the documentation: https://organice.200ok.ch/documentation.html#how_we_work_with_issues
- I understand this is a really tough feature
- I read that, as this is a long term issue. As of 2020-Jun, it is non-actionable. 
- As nobody will be working on it, issue will be closed, which is understandable
- I filed this so that even if this issued is closed, the closed issue can still be commented-upon if and when someone has made progress or has new information.
- nb. I am not a regular emacs user, I am also new to org-mode

**Comment by munen** [comment]
Hi @hgkamath 

Thank you for using organice :pray: 

It's not exactly what you're proposing, but I this is my workflow on how to get Org mode agenda items into my calendar (and hence on my phone with notifications): https://200ok.ch/posts/2022-02-13_integrating_org_mode_agenda_into_other_calendar_apps.html

Apart from that, we will be working on adding notifications to organice itself.

Adding a full sync between iCal and Org mode will likely never be part of organice. This is a feature that many people want in Emacs Org mode already and that nobody was willing to successfully provide in the last decades. For me personally, this is easy to understand:

1. Synching between disjunct protocols is hard
2. Bi-directional synching is even harder
3. There's different iCal implementations
4. Most importantly: Why would you want to have multiple tools for inputting the same kind of information? To me, that sounds like there's a problem somewhere else. What functionality is missing in tool A so that I want to do the same thing, but slightly different in tool B?

Addendum to 4: Just synching one into the other doesn't fall into 4. Exporting your Org agenda as ICS is something that makes sense, because it allows access to the same information from other tools and for people outside of the Org mode eco system. 

I hope any of that is of help. In any case, as you've previously stated, unless you plan on building a bi-directional sync in organice yourself, this issue will be closed. If you do want to build it, I'm absolutely happy to provide you pointers, do code reviews and help with testing!

Good luck and have fun :pray:

**Comment by hgkamath** [comment]
Presently, I don't think I can commit to doing it.  But, one day, if it feels like I wanna give it a try, I'll ping you, here. Thanks for offering to mentor.

- Thanks for the link to your blog post write-up. Yours is the latest at 2022-02-13 as of this writing.

- Notifications could be useful. There are notification that happen inside the PWA and notfications that are relayed to the smartphone/tablet OS. In either case, a disadvantage, that I could think of is, any app that needs a thread to awake and do something every now and then could make the pwa/app on the smartphone/tablet consume more battery. Apps that are not designed to stay in memory may get pushed to background and then put into sleep, thus delaying notification until foregrounded. I speculate that, a smartphone/tablet's calendar service is already designed to notify on events, and the OS may keep some small thread read to trigger/respond on a very-low-power-core even when smartphone/tablet is sleeping.

* >Why would you want to have multiple tools for inputting the same kind of information? 
  
  As you say, one could restrict one self to editing only in a single app (organice). One would have to forego the handy webmail/messaging features that automatically add appointment upon click through. 

* >Addendum to 4
  
  One directional ical-export and sync-ing is desirable in itself and could be a half-mile-milestone, which would also be useful to some extent.

* >...to provide you pointers
  
  Please do document all that you know, one never knows if someone would take it up.
 
Just documenting other links I encountered in the search of all this.
- Mike Zamansky Using Emacs - 26 - Google Calendar, Org Agenda https://cestlaz.github.io/posts/using-emacs-26-gcal/#.WIqBud9vGAk
- Rainer König in his youtube videos has a method: https://www.youtube.com/watch?v=cIzzjSaq2N8&amp;list=PLVtKhBrRV_ZkPnBtt_TD1Cs9PJlU0IIdE&amp;index=33
- Another attempt at org-mode syncing with android app : https://github.com/larkery/orgcal 
- Some collected reddit links
  - 2021 https://www.reddit.com/r/orgmode/comments/l2oxkd/orgmode_and_ical_how_to_accomplish/
  - 2020 https://www.reddit.com/r/orgmode/comments/j0zp0q/org_mode_and_google_calendar/
  - 2019 https://www.reddit.com/r/orgmode/comments/dca4ag/are_there_any_simple_alternatives_for_emacs/
  - 2019 https://www.reddit.com/r/orgmode/comments/argzvu/how_to_use_emacs_orgmode_on_android_and_ios/

**Comment by munen** [comment]
> Thank-you many many times more for creating organice.

You're very welcome. Thank you for contributing to FOSS🙏

> Presently, I don't think I can commit to doing it. But, one day, if it feels like I wanna give it a try, I'll ping you, here. Thanks for offering to mentor.

👍

> - Notifications could be useful. There are (a) those notifications
>   that happen inside an app/pwa and (b) those notifications that are
>   relayed to the smartphone/tablet OS. In either case, I think a
>   disadvantage exists.
>   - Any app that needs a thread to awake and do something every now
>     and then could make the app/pwa on the smartphone/tablet consume
>     more power and deplete battery faster.

Fortunately, this issue has been solved in mobile apps for iOS and Android years ago. I cannot remember the specific time, because I'm not a mobile app developer, but it's something like 8 years ago iirc.

TL;DR: Apps **cannot** run in the background. iOS and Android are not like other OSs in that regard. They only allow a specific subset of APIs to be called upon by apps that are not in the foreground.

> * >Why would you want to have multiple tools for inputting the same
>   kind of information?
>   
>   As you say, one could restrict one self to editing only in a single >   app (organice). One would have to forego the handy >   webmail/mailclient/messaging features that automatically add >   appointment upon click through.

You don't have to forgo that. You can totally add calendar invitations to Org mode files. Here's how I'm doing it in Emacs: https://github.com/munen/emacs.d/#calendar-invitations

> * >Addendum to 4
>   
>   One directional ical-export and one-way-sync-ing is a desirable
>   feature in-of-itself and could be a half-mile-milestone. There
>   could even be a way to filter entries by tag/heading and export to
>   different icals. Should RFE for one-directional-ical-export be
>   filed a different issue?
> 
> - If organice had a plugin system, RFEs such as these could be
>    implemented as plugins. That would allow users to mix and match,
>    make and publish minor plugin variations, and select what plugins
>    suits them. The RFE for a plugin system could also be filed as
>    another issue.

Sounds interesting. This proposal is a first. If you want to 

> - I noticed that the another software, logseq, a GUI outliner tool,
>   could read files in [logseq orgmode
>   syntax](https://docs.logseq.com/#/page/org%20mode). I think its
>   support is still very rudimentary. I didn't yet find much info if
>   it can actually do
>   [todo](https://docs.logseq.com/#/page/tasks%20%26%20todos)
>   management in the way org-mode does. It so far has a few todo
>   next/waiting/done primitives, fast growing plugin system, which at
>   present has calendar import, and also ability to do
>   [datalog](https://docs.logseq.com/#/page/advanced%20queries)
>   DSL-queries ([wiki](https://en.wikipedia.org/wiki/Datalog)) .

Yes, logseq is another tool that reads Org mode files. There are many and they do different things differently. Last time I checked logseq, I couldn't use it at all, because it didn't work well with my large Org files.

As for query language: Please note that organice has a powerful query syntax, as well. More about that in the [online help](https://organice.200ok.ch/sample) under Search.

---

Thank you for providing this RFE and for the additional pointers🙏 For the moment, as you proposed yourself, I'm going to close the issue.

When someone else is intent on implementing either ical integration or a plugin system, please feel free to re-open or comment on this issue.
